shadowsocks/config.go

```diff
@@ -0,0 +1,148 @@
+	// TLS certificate.
+	TrustedCertFingerprint []byte
+}
+
+// FetchConfigResponse encapsulates a response from an online config server.
+type FetchConfigResponse struct {
+	// Config is the parsed server response.
+	Config OnlineConfig
+	// HTTPStatusCode is the HTTP status code of the response.
+	HTTPStatusCode int
+	// RedirectURL is the Location header of a HTTP redirect response.
+	RedirectURL string
+}
+
+// OnlineConfig represents a SIP008 response from an online config server.
+type OnlineConfig struct {
+	Proxies []Config `json:"servers"`
+	Version int
+}
+
+// FetchConfig retrieves Shadowsocks proxy configurations per SIP008:
+// https://github.com/shadowsocks/shadowsocks-org/wiki/SIP008-Online-Configuration-Delivery
+//
+// Pins the trusted certificate when req.TrustedCertFingerprint is non-empty.
+// Sets the response's RedirectURL when the status code is a redirect.
+// Returns an error if req.URL is a non-HTTPS URL, if there is a connection
+// error to the server, or if parsing the configuration fails.
+func FetchConfig(req FetchConfigRequest) (*FetchConfigResponse, error) {
+	httpreq, err := http.NewRequest(req.Method, req.URL, nil)
+	if err != nil {
+		return nil, err
+	}
+	if httpreq.URL.Scheme != "https" {
+		return nil, errors.New("URL protocol must be HTTPs")
+	}
+
+	client := &http.Client{
+		// Do not follow redirects automatically, relay to the caller.
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+		Timeout: 30 * time.Second,
+	}
+
+	if req.TrustedCertFingerprint != nil && len(req.TrustedCertFingerprint) != 0 {
+		client.Transport = &http.Transport{
+			// Perform custom server certificate verification by pinning the
+			// trusted certificate fingerprint.
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify:    true,
+				VerifyPeerCertificate: makePinnedCertVerifier(req.TrustedCertFingerprint),
+			},
+		}
+	}
+
+	httpres, err := client.Do(httpreq)
+	if err != nil {
+		return nil, err
+	}
+
+	var res FetchConfigResponse
+	res.HTTPStatusCode = httpres.StatusCode
+	if res.HTTPStatusCode >= 300 && res.HTTPStatusCode < 400 {
+		// Redirect
+		res.RedirectURL = httpres.Header.Get("Location")
+		return &res, nil
+	} else if res.HTTPStatusCode > 400 {
+		// HTTP error
+		return &res, nil
+	}
+
+	// 2xx status code
+	defer httpres.Body.Close()
+	var config OnlineConfig
+	err = json.NewDecoder(httpres.Body).Decode(&config)
+	res.Config = config
+	return &res, err
+}
+
+type certVerifier func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
+
+// Verifies whether the pinned certificate SHA256 fingerprint,
+// trustedCertFingerprint, matches a fingerprint in the certificate chain,
+// regardless of the system's TLS certificate validation errors.
+func makePinnedCertVerifier(trustedCertFingerprint []byte) certVerifier {
+	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		for _, cert := range rawCerts {
```

Comment:
If `InsecureSkipVerify` is true, does `crypto/tls` still require that the certificate chain is internally consistent (i.e. each certificate signs the preceding one)?  If not, an adversary could simply copy-paste the true certificate into their cert chain to pass this check.

I see a few options:
1. Just check the fingerprint of `rawCerts[0]`.  I don't see why we would need to support chaining.
2. Scan `rawCerts` for a cert matching the fingerprint (`rootCert`), and do `leafCert.Verify` using `rawCerts` as intemediates and `rootCert` as the root CA.

Reply:
Works for me.